### PR TITLE
Added experimental custom RenderLayer support for Blocks

### PR DESCRIPTION
**Pull request by gudenau**
There was a comment that said it required a lot of hacks, when in reality it was only a pair of mixins.

There might be some edge cases where it doesn't work correctly, but it is in the experimental package for a reason.

### CODE REVIEW

File: src/main/java/ladysnake/satin/impl/BlockRenderLayerRegistryImplementation.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    
+    private final Set<RenderLayer> renderLayers = new HashSet<>();
+    private volatile boolean registryLocked = false;
+    
+    private BlockRenderLayerRegistryImplementation(){}
+    
+    @Override
+    public void registerRenderLayer(RenderLayer layer) {
+        if(registryLocked){
+            throw new RuntimeException(String.format(
```
Comment (Pyrofab) [inline src/main/java/ladysnake/satin/impl/BlockRenderLayerRegistryImplementation.java:19]:
```suggestion
            throw new IllegalStateException(String.format(
```

File: src/main/java/ladysnake/satin/impl/BlockRenderLayerRegistryImplementation.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package ladysnake.satin.impl;
+
+import java.util.HashSet;
+import java.util.Set;
+import ladysnake.satin.api.experimental.BlockRenderLayerRegistry;
+import net.minecraft.client.render.RenderLayer;
+
+public class BlockRenderLayerRegistryImplementation implements BlockRenderLayerRegistry {
```
Comment (Pyrofab) [inline src/main/java/ladysnake/satin/impl/BlockRenderLayerRegistryImplementation.java:8]: Should be called `BlockRenderLayerRegistryImpl`

File: src/main/java/ladysnake/satin/mixin/client/render/RenderLayerMixin.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        super(null, null, null);
+    }
+    
+    @Redirect(
+        method = "getBlockLayers",
+        at = @At(
+            value = "INVOKE",
+            target = "Lcom/google/common/collect/ImmutableList;of(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Lcom/google/common/collect/ImmutableList;"
+        )
+    )
```
Comment (Pyrofab) [inline src/main/java/ladysnake/satin/mixin/client/render/RenderLayerMixin.java:25]: Not fond of the redirect. I would instead inject at return, then `cir.setReturnValue(ImmutableList.builder().addAll(cir.getReturnValue()).addAll(BlockRenderLayerRegistryImpl.INSTANCE.getLayers()))`

File: src/main/java/ladysnake/satin/mixin/client/render/WorldRendererMixin.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+public abstract class WorldRendererMixin implements SynchronousResourceReloadListener, AutoCloseable{
+    @Shadow protected abstract void renderLayer(RenderLayer layer, MatrixStack matrix, double x, double y, double z);
+    
+    @Redirect(
+        method = "render",
+        at = @At(
+            value = "INVOKE",
+            target = "Lnet/minecraft/client/render/WorldRenderer;renderLayer(Lnet/minecraft/client/render/RenderLayer;Lnet/minecraft/client/util/math/MatrixStack;DDD)V",
+            ordinal = 2
+        ),
```
Comment (Pyrofab) [inline src/main/java/ladysnake/satin/mixin/client/render/WorldRendererMixin.java:24]: You do not need a redirect, you can just capture the variables as they are computed very early in the method.

**Review by Pyrofab — CHANGES_REQUESTED** [review]
The "lots of hacks" comment was based on [this comment on a FAPI issue](https://github.com/FabricMC/fabric/issues/642#issuecomment-635019528). I am still not sure it's a good idea, but if it works it works. I think the code could still be improved though.

**Comment by gudenau** [comment]
Alright, this should have the suggested changes now.

File: src/main/java/ladysnake/satin/mixin/client/render/WorldRendererMixin.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                value = "INVOKE",
+                target = "Lnet/minecraft/client/render/SkyProperties;isDarkened()Z"
+            )
+        ),
+        locals = LocalCapture.CAPTURE_FAILHARD
+    )
+    private void renderCustom(
+        MatrixStack matrix, float delta, long time, boolean renderOutlines, Camera camera, GameRenderer gameRenderer, LightmapTextureManager lightManager, Matrix4f frustumMatrix,
+        CallbackInfo info,
+        Profiler profiler, Vec3d vec3d, double x, double y, double z, Matrix4f matrix4f2, boolean bl, Frustum frustum2, float g, boolean bl2, int i, int j, long n, long o, long p, long q, long r
```
Comment (Pyrofab) [inline src/main/java/ladysnake/satin/mixin/client/render/WorldRendererMixin.java:50]:
```suggestion
        Profiler profiler, Vec3d vec3d, double x, double y, double z
```
you don't have to capture *every* variable, you can stop once you got the ones you need.

File: src/main/java/ladysnake/satin/mixin/client/render/RenderLayerMixin.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    }
+    
+    @Inject(
+        method = "getBlockLayers",
+        at = @At("RETURN"),
+        cancellable = true
+    )
+    private static void getBlockLayers(CallbackInfoReturnable<ImmutableList<Object>> info) {
+        info.setReturnValue(
+            new ImmutableList.Builder<>()
```
Comment (Pyrofab) [inline src/main/java/ladysnake/satin/mixin/client/render/RenderLayerMixin.java:25]:
```suggestion
            ImmutableList.builder()
```

**Review by Pyrofab — CHANGES_REQUESTED** [review]
Better, just a few style issues left. I would also like @grondag's opinion on this.

File: src/main/java/ladysnake/satin/mixin/client/render/RenderLayerMixin.java (hunk 6 above)
Comment (gudenau) [reply src/main/java/ladysnake/satin/mixin/client/render/RenderLayerMixin.java:25]: Oops.

File: src/main/java/ladysnake/satin/mixin/client/render/WorldRendererMixin.java (hunk 5 above)
Comment (gudenau) [reply src/main/java/ladysnake/satin/mixin/client/render/WorldRendererMixin.java:50]: I did not know that.

**Review by Pyrofab — APPROVED** [review]

**Comment by grondag** [comment]
This is something I've been wrestling with for Canvas recently, so I have some thoughts:

In the end I don't think exposing `RenderLayer` directly is a good idea, for several reasons:

1) It's likely to change significantly in subsequent MC versions
2) Related to 1, it doesn't have a public, well-defined or well-supported interface - it's a raw implementation.
3) It's an obstacle to optimization or alternate rendering pipelines where similar effects can be achieved with fewer draw calls or in conjunction with an improved lighting model.
4) The timing of draws and configuration of targets can affect or be affected by other things that happen in the pipeline.  

To elaborate on this last point, some layers can be deferred and drawn together (but might have to happen before or after others), while others might have to be drawn immediately.  For vanilla layers it manages to work via implicit usage assumptions by Mojang and the hardcoded sequence of events in WorldRenderer. For custom layers there are no semantics or contracts - it will be chaos - or more likely you have to assume/hope the immediate buffer is truly empty and draw it after every use. 

Where my mind is headed is to extend the Fabric `RenderMaterial` and corresponding builder to expose the various effects currently enabled by RenderLayer and provide a default implementation based on RenderLayer. For that to be useful, there would need to be an alternate to VertexConsumer analogous to QuadEmitter that allows for material assignment.  (And it would have to allow other primitive types beyond quads.)

Mod authors would also need a way to access this facility in different contexts, like entity rendering, BE rendering, particles, etc.  Not yet sure how best to provide this - could be an interface implemented by those classes to be called at render time, registration of handlers to be called in the same way (but probably too slow to look them up), or static access as needed.  Or perhaps some combination or alternative I hadn't considered.

An API like this has several benefits:

1) Breaking changes by Mojang can be handled once, behind the interface. Mods using the API should see relatively fewer breaks. 
2) Clear separation of interface and implementation.  A mod like Canvas or Sodium can change the implementation of material drawing to improve performance and/or aesthetics without breaking mods.
3) Canvas, Sodium or other renderers can expand the available choices via shaders or additional parameters.
4) The API can be easier to use by exposing logical choices vs GL implementation details.
5) The API can accept additional metadata about render intent so that how and when drawing happens is never ambiguous.

To illustrate how different the implementation can be, consider that Canvas will achieve the effects of RenderLayer via a combination of vertex data and uniforms interpreted by shaders instead of GL draw state alone.  Generally, GL state is the _last_ thing Canvas wants to change because it can stall the pipeline.  Canvas may also combine buffers and/or draw calls when uniforms/vertex data alone are sufficient.  (Cutout and mip-mapping, for example, are controlled by bit flags in vertex data.)

As another example, Canvas also intends to support Physically-Based Rendering texture maps for entities and block entities, which means it must modify and control the draw state and draw calls for those contexts.  Doing this well and reliably with any random RenderLayer it encounters would be impractical. 

So, that's my rough notion.  I would certainly be open to a collaboration if you are interested. I had planned to expose the API in FREX with a default implementation for non-Canvas users, or perhaps as a standalone mod. It may eventually be suitable for a Fabric PR but that would not be fast.

I appreciate being asked for input but it is of course entirely your choice.  If you decide to add this feature in Satin, I'll not complain, but be aware that Canvas will almost certainly not be able to offer compatibility. I can't speak for Jelly, but I would imagine the challenges are similar for Sodium.

**Comment by grondag** [comment]
I want to elaborate a bit on the performance concern: adding block render layers in terrain scales poorly if done simply by adding more buffers and render layers.  As the number of layers expands, for each render region there can be more buffer bindings, more GL state changes, more draw calls, etc.  

It doesn't scale for translucency _at all_ assuming one works through the complications of sorting quads across more than one layer and then somehow orchestrating the sequence of state changes and draw calls to maintain correct ordering.  I've tried it before in the hope it might be tolerable, and can [confirm performance sucks](https://github.com/grondag/canvas/issues/31).

Canvas and Sodium both employ complicated strategies to reduce the impact of draw calls and state changes.  Sodium does it with Vanilla as the baseline, with the aim of improving overall performance.  Canvas does it to minimize the impact of custom render effects.

For example, Canvas renders all solid and cutout terrain from a single buffer, and for many materials in a single pass.  Effects with different shaders currently require separate draw calls but that too will be changing.  It also reduces matrix state changes and overall bandwidth usage by altering vertex formats and coordinates.

That said, a simple implementation like the one in this PR will probably work well enough for one-off render layers added by a few mods for blocks that appear infrequently.  If you proceed, I suggest you advise authors to create and use custom layers sparingly, and that you restrict custom translucent layers altogether.  

As noted previously, I won't be able to ensure compatibility with Canvas and I will eventually be introducing an extension of FRAPI that covers the same use cases  per my plans laid out above.

**Comment by Pyrofab** [comment]
Considering the use case for this PR (arbitrary shaders for blocks) is fundamentally incompatible with a generic abstraction to be implemented by Canvas and Sodium, I think I will merge this PR and work out automatic deactivation when an alternative renderer is found.

**Pull request merged**